source/fuzz/fuzzer_pass_replace_loads_stores_with_copy_memories.cpp

```diff
@@ -0,0 +1,108 @@
+// Copyright (c) 2020 Google LLC
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#include "source/fuzz/fuzzer_pass_replace_loads_stores_with_copy_memories.h"
+
+#include "source/fuzz/fuzzer_util.h"
+#include "source/fuzz/instruction_descriptor.h"
+#include "source/fuzz/transformation_replace_load_store_with_copy_memory.h"
+#include "source/opt/instruction.h"
+
+namespace spvtools {
+namespace fuzz {
+
+FuzzerPassReplaceLoadsStoresWithCopyMemories::
+    FuzzerPassReplaceLoadsStoresWithCopyMemories(
+        opt::IRContext* ir_context,
+        TransformationContext* transformation_context,
+        FuzzerContext* fuzzer_context,
+        protobufs::TransformationSequence* transformations)
+    : FuzzerPass(ir_context, transformation_context, fuzzer_context,
+                 transformations) {}
+
+FuzzerPassReplaceLoadsStoresWithCopyMemories::
+    ~FuzzerPassReplaceLoadsStoresWithCopyMemories() = default;
+
+void FuzzerPassReplaceLoadsStoresWithCopyMemories::Apply() {
+  // A vector of matching OpLoad and OpStore instructions.
+  std::vector<std::pair<opt::Instruction*, opt::Instruction*>>
+      op_load_store_pairs;
+
+  for (auto& function : *GetIRContext()->module()) {
+    for (auto& block : function) {
+      // A hash map storing potential OpLoad instructions.
+      std::unordered_map<uint32_t, opt::Instruction*> current_op_loads;
+      for (auto& instruction : block) {
+        // Add an potential OpLoad instruction.
+        if (instruction.opcode() == SpvOpLoad) {
+          current_op_loads[instruction.result_id()] = &instruction;
+        } else if (instruction.opcode() == SpvOpStore) {
+          if (current_op_loads.find(instruction.GetSingleWordOperand(1)) !=
+              current_op_loads.end()) {
+            // We have found the matching OpLoad instruction to the current
+            // OpStore instruction.
+            op_load_store_pairs.push_back(std::make_pair(
+                current_op_loads[instruction.GetSingleWordOperand(1)],
+                &instruction));
+          }
+          current_op_loads.clear();
```

Comment:
This looks good. I was thinking of something more like this:

```
if opcode == SpvOpLoad:
  ...
else if opcode == SpvOpStore
  // add pair.

if IsMemoryWritingOpcode(opcode):
  current_op_loads.clear()
else if IsMemoryBarrierOpcode(opcode):
  for load in current_op_loads:
    storage_class = ...load...
    if !IsStorageClassSafeAcrossMemoryBarriers(storage_class):
      erase load
```

Where `IsMemoryWritingOpcode` would return true for `SpvOpStore`, `SpvOpCopyMemory`, `SpvOpCopyMemorySized`, and all atomic instructions **except** for `OpAtomicLoad` (and we might exclude other instructions some day). 

Then in IsApplicable, you could re-use most of this from within your loop:

```
if IsMemoryWritingOpcode(opcode):
  return false
else if IsMemoryBarrierOpcode(opcode) and !IsStorageClassSafeAcrossMemoryBarriers(storage_class):
  return false
```

Reply:
Yes, this solution looks much cleaner and it doesn't reuse code. Thanks. I'll implement it.